lib/stdlib/test/shell_SUITE.erl

```diff
@@ -256,6 +256,8 @@ restricted_local(Config) when is_list(Config) ->
                   local_allowed(_,_,State) ->
                       {false,State}.
 
+                  non_local_allowed({erlang,raise},[_, _, _],State) ->
+                      {true,State};
```

Comment:
@bjorng this is a potential backwards incompatible change. Now all errors are raised through the evaluator, which means a restricted shell needs to allow them by allowing `erlang:raise/3` as above. Not allowing this means that accessing an undefined variable now says the operation is not restricted instead of raising.

This can be seen as a feature, but it can also be breaking. Other options are:

1. Have `erl_eval:raise/3` instead, with the `Anno`, `Reason`, and `Stacktrace`, to work as a clearer flag to check in the restricted shell, although ultimately it doesn't change much

2. Automatically raise for `erlang:raise/3`, instead of asking the restricted shell

If any of those changes are not desired, then the only option is to add a new callback to `erl_eval`.

Reply:
> Not allowing this means that accessing an undefined variable now says the operation is not restricted instead of raising.

Did you mean?

> ... is restricted instead of raising.

Also, it seems that accessing undefined variables will usually be catched by `erl_lint`:

```
Restricted Eshell V12.2  (abort with ^G)
1> Undefined.
* 1:1: variable 'Undefined' is unbound
```

For other types of errors, such various record errors and failure to match, there will be an ugly error message:

```
2> a = b.
** exception exit: restricted shell does not allow erlang:raise(error,
                                                                {badmatch,b},
                                                                [{erl_eval,expr,6,[{file,"erl_eval.erl"},{line,470}]},
                                                                 {shell,exprs,7,[{file,"shell.erl"},{line,689}]},
                                                                 {shell,eval_exprs,7,[{file,"shell.erl"},{line,645}]},
                                                                 {shell,restricted_eval_loop,4,
                                                                        [{file,"shell.erl"},{line,639}]}])
```

In OTP 24, the message would be:

```
1> a = b.
** exception error: no match of right hand side value b
```

I think that this change is acceptable from a backward compatibility perspective. Everything will still work if a restrictive shell does not let through `erlang:raise/3` but error messages will be harder to read.

@KennethL and @jhogberg, what do you think?

Reply:
Additional idea after discussion with @jhogberg: When an exception is thrown because the restricted shell does not allow it, the catcher of the exception (the module `erl_error`, I think), should be able to figure out that the original call originated from `erl_eval` and reconstruct the original error message.

Reply:
@bjorng do you mean we could look at the stacktrace top and see if it comes from `erl_eval`?

Reply:
Yes.

Reply:
Doing that or keeping it as is both work for me. Let me know if you have a preference. :)

Reply:
Perhaps I would keep it as is, to make it clear for those in restricted shells that they may want to handle it, and consider adding the mentioned fallback based on feedback.

Reply:
> Perhaps I would keep it as is, to make it clear for those in restricted shells that they may want to handle it, and consider adding the mentioned fallback based on feedback.

OK. Let's keep it and wait for feedback.

We'll need an update of the documentation for restricted shell.